src/widgets/scrollbar.rs

```diff
@@ -0,0 +1,698 @@
+    }
+
+    /// Decrements the scroll position by one, ensuring it doesn't go below zero.
+    pub fn prev(&mut self) {
+        self.position = self.position.saturating_sub(1);
+    }
+
+    /// Increments the scroll position by one, ensuring it doesn't exceed the length of the content.
+    pub fn next(&mut self) {
+        self.position = self
+            .position
+            .saturating_add(1)
+            .clamp(0, self.content_length.saturating_sub(1))
+    }
+
+    /// Sets the scroll position to the start of the scrollable content.
+    pub fn first(&mut self) {
+        self.position = 0;
+    }
+
+    /// Sets the scroll position to the end of the scrollable content.
+    pub fn last(&mut self) {
+        self.position = self.content_length.saturating_sub(1)
+    }
+
+    /// Changes the scroll position based on the provided ScrollDirection.
+    pub fn scroll(&mut self, direction: ScrollDirection) {
+        match direction {
+            ScrollDirection::Forward => {
+                self.next();
+            }
+            ScrollDirection::Backward => {
+                self.prev();
+            }
+        }
+    }
+}
+
+/// Scrollbar Orientation
+#[derive(Default, Debug, Clone)]
+pub enum ScrollbarOrientation {
+    #[default]
+    Vertical,
+    Horizontal,
+}
+
+/// Scrollbar widget for tui-rs library.
+///
+/// This widget can be used to display a scrollbar in a terminal user interface.
+#[derive(Debug, Clone)]
+pub struct Scrollbar<'a> {
+    pub orientation: ScrollbarOrientation,
+    pub thumb_style: Style,
+    pub thumb_symbol: &'a str,
+    pub track_style: Style,
+    pub track_symbol: &'a str,
+    pub begin_symbol: Option<&'a str>,
+    pub begin_style: Style,
+    pub end_symbol: Option<&'a str>,
+    pub end_style: Style,
+    pub mirror: bool,
+}
+
+impl<'a> Default for Scrollbar<'a> {
+    fn default() -> Self {
+        Self {
+            orientation: ScrollbarOrientation::default(),
+            thumb_symbol: DOUBLE_VERTICAL.thumb,
+            thumb_style: Style::default(),
+            track_symbol: DOUBLE_VERTICAL.track,
+            track_style: Style::default(),
+            begin_symbol: Some(DOUBLE_VERTICAL.begin),
+            begin_style: Style::default(),
+            end_symbol: Some(DOUBLE_VERTICAL.end),
+            end_style: Style::default(),
+            mirror: false,
+        }
+    }
+}
+
+impl<'a> Scrollbar<'a> {
+    pub fn new(orientation: ScrollbarOrientation) -> Self {
+        Self::default().orientation(orientation)
+    }
+
+    /// Sets the orientation of the scrollbar.
+    /// Resets the symbols to [`DOUBLE_VERTICAL`] or [`DOUBLE_HORIZONTAL`] based on orientation
+    pub fn orientation(mut self, orientation: ScrollbarOrientation) -> Self {
+        self.orientation = orientation.clone();
+        self.symbol(match orientation {
+            ScrollbarOrientation::Vertical => DOUBLE_VERTICAL,
+            ScrollbarOrientation::Horizontal => DOUBLE_HORIZONTAL,
+        })
+    }
```

Comment:
> `ScrollBar::default().orientation(Orientation::Horizontal) != ScrollBar::new(Orientation::Horizontal)`

Why do you say that this is the case? Aren't they identical?

I would say that majority of the users are going to want to just choose between a Vertical or Horizontal scrollbar, so we want to make that "path" easy to use. Which is why I think setting orientation should change the symbols used.

Reply:
Yep makes sense - I just misread the way the interaction between new and default worked.
I'm used to seeing new actually returning Self { ... } and assumed wrongly that it would be setting the orientation directly on the new struct, but it's actually creating it in default() and then setting the orientation with the builder method. This is fine I think.